odl/discr/discr_mappings.py

```diff
@@ -662,33 +642,94 @@ def __call__(self, xi, outp=None):
             raise ValueError('number of vectors in `xi` is {} instead of {}, '
                              'the grid dimension.'.format(xi.shape[1],
                                                           len(self.grid)))
-        ntotal = np.prod(np.broadcast_arrays(*xi)[0].shape)
-        if outp is not None:
-            if not isinstance(outp, np.ndarray):
-                raise TypeError('`outp` {!r} not a `numpy.ndarray` '
-                                'instance.'.format(outp))
-            if outp.shape != (ntotal,):
+
+        if len(xi) == 1:
+            ntotal = xi[0].size
+        else:
+            ntotal = np.prod(np.broadcast(*xi).shape)
+        if out is not None:
+            if not isinstance(out, np.ndarray):
+                raise TypeError('`out` {!r} not a `numpy.ndarray` '
+                                'instance.'.format(out))
+            if out.shape != (ntotal,):
                 raise ValueError('Output shape {} not equal to (n,), where '
                                  'n={} is the total number of evaluation '
-                                 'points.'.format(outp.shape, ntotal))
+                                 'points.'.format(out.shape, ntotal))
 
         indices, norm_distances = self._find_indices(xi)
-        return self._evaluate_nearest(indices, norm_distances, outp)
+        return self._evaluate(indices, norm_distances, out)
 
-    def _evaluate_nearest(self, indices, norm_distances, outp=None):
-        """Evaluate nearest interpolation.
 
-        Modified for in-place evaluation.
-        """
+class _NearestPointwiseInterpolator(_PointwiseInterpolator):
+
+    """Nearest neighbor interpolator for point arrays.
+
+    The code is adapted from SciPy's `RegularGridInterpolator
+    <http://docs.scipy.org/doc/scipy/reference/generated/\
+scipy.interpolate.RegularGridInterpolator.html>`_ class.
+    """
+
+    def _evaluate(self, indices, norm_distances, out=None):
+        """Evaluate nearest interpolation. Modified for in-place."""
         idx_res = []
         for i, yi in zip(indices, norm_distances):
+            # TODO: adapt here for the different cases [a, b), (a, b] or (a, b)
             idx_res.append(np.where(yi <= .5, i, i + 1))
-        if outp is not None:
-            outp[:] = self.values[idx_res]
-            return outp
+        if out is not None:
+            out[:] = self.values[idx_res]
+            return out
         else:
             return self.values[idx_res]
 
+
+class _NearestMeshgridInterpolator(_MeshgridInterpolator,
+                                   _NearestPointwiseInterpolator):
+
+    """Nearest neighbor interpolator for point meshgrids.
+
+    The code is adapted from SciPy's `RegularGridInterpolator
```

Comment:
I had to adapt things internally, e.g. remove the cast to `float` in the nearest interpolation, flip axes. Possibly more changes to come. None of the classes is mere copy paste.